library/api.inc

```diff
@@ -27,11 +27,48 @@ function formFooter()
     <?php
 }
 
-function formSubmit($tableName, $values, $id, $authorized = "0")
+/**
+ * Original Query: "insert into " . escape_table_name($tableName) . " set " . $attendant_type . "='" . add_escape_custom($_SESSION['pid']) . "',groupname='" . add_escape_custom($_SESSION['authProvider']) . "',user='" . add_escape_custom($_SESSION['authUser']) . "',authorized='" . add_escape_custom($authorized) . "',activity=1, date = NOW(),";
+ *
+ * @param  $tableName
+ * @param  $attendant_type
+ * @return string
+ */
+function buildInsertFormResultsQuery($tableName, $params)
 {
-        global $attendant_type;
+    $escapedTableName = escape_table_name($tableName);
+    $boundColumns = implode('= ?,', array_keys($params)) . '= ?,';
+    return <<<QUERY
+INSERT INTO {$escapedTableName} SET date = NOW(), {$boundColumns};
+QUERY;
+}
 
-    $sql = "insert into " . escape_table_name($tableName) . " set " .  $attendant_type . "='".add_escape_custom($_SESSION['pid'])."',groupname='".add_escape_custom($_SESSION['authProvider'])."',user='".add_escape_custom($_SESSION['authUser'])."',authorized='".add_escape_custom($authorized)."',activity=1, date = NOW(),";
+/**
+ * original update " . escape_table_name($tableName) . " set pid ='" . add_escape_custom($_SESSION['pid']) . "',groupname='" . add_escape_custom($_SESSION['authProvider']) . "',user='" . add_escape_custom($_SESSION['authUser']) . "',authorized='" . add_escape_custom($authorized) . "',activity=1, date = NOW(),
+ *
+ * @param  $tableName
+ * @param  $inputParams
+ * @return string
+ */
+function buildUpdateFormResultsQuery($tableName, $inputParams)
+{
+    $escapedTableName = escape_table_name($tableName);
+    $boundColumns = implode('= ?,', array_keys($inputParams)) . '= ?,';
+    return <<<QUERY
+UPDATE {$escapedTableName} SET date = NOW(), {$boundColumns} WHERE id=?;
+QUERY;
+}
```

Comment:
bookmark to re-analyze above after complete review of this script

Reply:
ok, had a chance to analyze this. First, this is really neat code to see and review. And these are definitely some of the more complicated sql queries that you took on. The main issue with sql queries is that it's best not to hide them behind functions like this or else sql-injection vectors may get through. For example, the $attendant_type column name needs some attention, however it may get missed in this code. And if devs build on these functions, then sql-injections vectors may get in. I just converted this script here as an example:
https://github.com/openemr/openemr/pull/2120/commits/b46f2f0bcbb3c9ad41518af858d72a8adf9364d1
(btw, note how I used your more modern way of array building, which looks much more clean)

Reply:
You're right. I usually use unit tests to validate something like this (or an ORM, or a Query Builder) - I converted it before I realized unit tests weren't present. 

Might I suggest a follow-up issue to make this more transparent - or is this a blocker?

Reply:
For this script, we could just probably go with the example script at:
https://github.com/openemr/openemr/pull/2120
(not fancy, but gets job done and minimal chance of bugs)